sql/src/main/java/io/crate/analyze/validator/Validator.java

```diff
@@ -0,0 +1,171 @@
+            if (symbol.valueType() != DataTypes.OBJECT && !isObjectArray(symbol.valueType())) {
+                return visitSymbol(symbol, context);
+            } else {
+                return new Function() {
+                    @Nullable
+                    @Override
+                    public Object apply(@Nullable Object input) {
+                        Map<String, Object> obj = (Map<String, Object>) input;
+                        normalizeObject(obj, symbol);
+                        return obj;
+                    }
+                };
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> normalizeObject(Map<String, Object> value, Reference ref) {
+        for (Map.Entry<String, Object> entry : value.entrySet()) {
+            ColumnIdent nestedIdent = ColumnIdent.getChild(ref.ident().columnIdent(), entry.getKey());
+            TableInfo tableInfo = referenceInfos.getTableInfoUnsafe(ref.info().ident().tableIdent());
+            ReferenceInfo nestedInfo = tableInfo.getReferenceInfo(nestedIdent);
+            if (nestedInfo == null) {
+                if (ref.info().columnPolicy() == ColumnPolicy.IGNORED) {
+                    continue;
+                }
+                DynamicReference dynamicReference = tableInfo.getDynamic(nestedIdent, false);
+                if (dynamicReference == null) {
+                    throw new ColumnUnknownException(nestedIdent.sqlFqn());
+                }
+                DataType type = DataTypes.guessType(entry.getValue(), false);
+                if (type == null) {
+                    throw new ColumnValidationException(ref.info().ident().columnIdent().sqlFqn(), "Invalid value");
+                }
+                dynamicReference.valueType(type);
+                nestedInfo = dynamicReference.info();
+            } else {
+                if (entry.getValue() == null) {
+                    continue;
+                }
+            }
+            if (nestedInfo.type() == DataTypes.OBJECT && entry.getValue() instanceof Map) {
+                value.put(entry.getKey(), normalizeObject((Map<String, Object>) entry.getValue(), new Reference(nestedInfo)));
+            } else if (isObjectArray(nestedInfo.type()) && entry.getValue() instanceof Object[]) {
+                value.put(entry.getKey(), normalizeObjectArrayValue((Object[])entry.getValue(), nestedInfo));
+            } else {
+                value.put(entry.getKey(), normalizePrimitiveValue(entry.getValue(), nestedInfo));
+            }
+        }
+        return value;
+    }
+
+    private boolean isObjectArray(DataType type) {
+        return type.id() == ArrayType.ID && ((ArrayType)type).innerType().id() == ObjectType.ID;
+    }
+
+    private Object normalizePrimitiveValue(Object primitiveValue, ReferenceInfo info) {
+        try {
+            if (info.type().equals(DataTypes.STRING) && primitiveValue instanceof String) {
+                return primitiveValue;
```

Comment:
as we always convert strings to byterefs, this condition should be removed afaik